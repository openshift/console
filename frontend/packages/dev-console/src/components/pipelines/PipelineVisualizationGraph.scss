$border-color: var(--pf-global--BorderColor--light-100);
$gutter: 1.8em;

.odc-pipeline-vis-graph {
  background: var(--pf-global--BackgroundColor--100);
  padding: var(--pf-global--spacer--md);
  margin: var(--pf-global--spacer--md) 0;
  height: 20em;
  overflow: auto;
  position: relative;
  -webkit-overflow-scrolling: touch;

  // reset
  &__stage-column {
    list-style: none;
    padding: 0;
    margin: 0;
  }

  &__stages {
    display: inline-flex;
  }

  &__stage {
    margin: 0 2em;

    &:not(:first-child) {
      & .odc-pipeline-vis-task:first-child {
        &::before {
          content: '';
          position: absolute;
          top: 1em;

          border-top: 1px solid $border-color;
          width: $gutter * 1.7 + 1;
          height: 0;
        }
      }
    }
  }

  &__stage.is-parallel {
    & .odc-pipeline-vis-task {
      &:first-child {
        &::before,
        &::after {
          content: '';
          position: absolute;
          top: 1em;
          border-top: 1px solid $border-color;
          width: $gutter * 1.7 + 1;
          height: 0;
        }
        &::before {
          left: 0;
          transform: translateX(-100%);
        }
        &::after {
          right: 0;
          transform: translateX(100%);
        }
      }
    }
    &:last-child {
      & .odc-pipeline-vis-task:first-child {
        &::after {
          content: '';
          width: $gutter / 2 + 0.05;
        }
      }
    }
  }
  &__stage:not(.is-parallel) {
    & .odc-pipeline-vis-task:first-child {
      &::before {
        left: 0;
        transform: translateX(-100%);
      }
    }
  }
}
