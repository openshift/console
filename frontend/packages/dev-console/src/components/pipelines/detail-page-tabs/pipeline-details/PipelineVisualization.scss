.odc-pipeline-visualization {
  margin-bottom: var(--pf-global--spacer--md);
}
