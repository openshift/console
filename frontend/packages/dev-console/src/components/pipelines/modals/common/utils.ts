import * as _ from 'lodash';
import { getRandomChars } from '@console/shared';
import { SecretKind } from '@console/internal/module/k8s';
import { SecretType } from '@console/internal/components/secrets/create-secret';
import {
  Pipeline,
  PipelineResource,
  PipelineRun,
  PipelineRunInlineResource,
  PipelineRunInlineResourceParam,
  PipelineRunReferenceResource,
  PipelineRunResource,
} from '../../../../utils/pipeline-augment';
import { PipelineRunModel } from '../../../../models';
import { getPipelineRunParams, getPipelineRunWorkspaces } from '../../../../utils/pipeline-utils';
import { SecretAnnotationId } from '../../const';
import { CREATE_PIPELINE_RESOURCE, initialResourceFormValues } from './const';
import { CommonPipelineModalFormikValues, PipelineModalFormResource } from './types';

/**
 * Migrates a PipelineRun from one version to another to support auto-upgrades with old (and invalid) PipelineRuns.
 *
 * Note: Each check within this method should be driven by the apiVersion number if the API is properly up-versioned
 * for these breaking changes. (should be done moving from 0.10.x forward)
 */
export const migratePipelineRun = (pipelineRun: PipelineRun): PipelineRun => {
  let newPipelineRun = pipelineRun;

  const serviceAccountPath = 'spec.serviceAccount';
  if (_.has(newPipelineRun, serviceAccountPath)) {
    // .spec.serviceAccount was removed for .spec.serviceAccountName in 0.9.x
    // Note: apiVersion was not updated for this change and thus we cannot gate this change behind a version number
    const serviceAccountName = _.get(newPipelineRun, serviceAccountPath);
    newPipelineRun = _.omit(newPipelineRun, [serviceAccountPath]);
    newPipelineRun = _.merge(newPipelineRun, {
      spec: {
        serviceAccountName,
      },
    });
  }

  return newPipelineRun;
};

export const getPipelineRunData = (
  pipeline: Pipeline = null,
  latestRun?: PipelineRun,
): PipelineRun => {
  if (!pipeline && !latestRun) {
    // eslint-disable-next-line no-console
    console.error('Missing parameters, unable to create new PipelineRun');
    return null;
  }

  const pipelineName = pipeline ? pipeline.metadata.name : latestRun.spec.pipelineRef.name;

  const resources = latestRun?.spec.resources;
  const workspaces = latestRun?.spec.workspaces;

  const latestRunParams = latestRun?.spec.params;
  const pipelineParams = pipeline?.spec.params;
  const params = latestRunParams || getPipelineRunParams(pipelineParams);

  const newPipelineRun = {
    apiVersion: pipeline ? pipeline.apiVersion : latestRun.apiVersion,
    kind: PipelineRunModel.kind,
    metadata: {
      name: `${pipelineName}-${getRandomChars(6)}`,
      namespace: pipeline ? pipeline.metadata.namespace : latestRun.metadata.namespace,
      labels: _.merge({}, pipeline?.metadata?.labels, latestRun?.metadata?.labels, {
        'tekton.dev/pipeline': pipelineName,
      }),
    },
    spec: {
      ...(latestRun?.spec || {}),
      pipelineRef: {
        name: pipelineName,
      },
      resources,
      ...(params && { params }),
      workspaces,
      status: null,
    },
  };
  return migratePipelineRun(newPipelineRun);
};

export const convertPipelineToModalData = (
  pipeline: Pipeline,
  alwaysCreateResources: boolean = false,
): CommonPipelineModalFormikValues => {
  const {
    metadata: { namespace },
    spec: { params, resources },
  } = pipeline;

  return {
    namespace,
    parameters: params || [],
    resources: (resources || []).map((resource: PipelineResource) => ({
      name: resource.name,
      selection: alwaysCreateResources ? CREATE_PIPELINE_RESOURCE : null,
      data: {
        ...initialResourceFormValues[resource.type],
        type: resource.type,
      },
    })),
  };
};

export const convertMapToNameValueArray = (map: {
  [key: string]: any;
}): PipelineRunInlineResourceParam[] => {
  return Object.keys(map).map((name) => {
    const value = map[name];
    return { name, value };
  });
};

const convertResources = (resource: PipelineModalFormResource): PipelineRunResource => {
  if (resource.selection === CREATE_PIPELINE_RESOURCE) {
    return {
      name: resource.name,
      resourceSpec: {
        params: convertMapToNameValueArray(resource.data.params),
        type: resource.data.type,
      },
    } as PipelineRunInlineResource;
  }

  return {
    name: resource.name,
    resourceRef: {
      name: resource.selection,
    },
  } as PipelineRunReferenceResource;
};

export const getPipelineRunFromForm = (
  pipeline: Pipeline,
  formValues: CommonPipelineModalFormikValues,
  labels?: { [key: string]: string },
) => {
  const { parameters, resources, workspaces } = formValues;

  const pipelineRunData: PipelineRun = {
    metadata: {
      labels,
    },
    spec: {
      pipelineRef: {
        name: pipeline.metadata.name,
      },
      params: getPipelineRunParams(parameters),
      resources: resources.map(convertResources),
      workspaces: getPipelineRunWorkspaces(workspaces),
    },
  };
  return getPipelineRunData(pipeline, pipelineRunData);
};

export const secretFormDefaultValues = {
  secretName: '',
  annotations: { key: SecretAnnotationId.Image, value: '' },
  type: SecretType.dockerconfigjson,
  formData: {},
};

export const isDuplicate = (secrets: SecretKind[], newSecretName: string): boolean => {
  return !!_.find(secrets, (obj) => obj.metadata.name === newSecretName);
};
