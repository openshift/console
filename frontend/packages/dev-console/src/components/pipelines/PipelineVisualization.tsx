import * as React from 'react';
import { K8sResourceKind } from '@console/internal/module/k8s';
import { getPipelineTasks } from '../../utils/pipeline-utils';
import { PipelineVisualizationGraph } from './PipelineVisualizationGraph';

export interface PipelineVisualizationProps {
  pipeline?: K8sResourceKind;
}

export const PipelineVisualization: React.FC<PipelineVisualizationProps> = ({ pipeline }) => (
  <PipelineVisualizationGraph
    namespace={pipeline.metadata.namespace}
    graph={getPipelineTasks(pipeline)}
  />
);
