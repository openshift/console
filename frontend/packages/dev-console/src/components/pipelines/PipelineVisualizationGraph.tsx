import * as React from 'react';
import * as cx from 'classnames';
import { ChevronCircleRightIcon } from '@patternfly/react-icons';
import { PipelineVisualizationTask } from './PipelineVisualizationTask';
import { PipelineVisualizationTaskItem } from '../../utils/pipeline-utils';

import './PipelineVisualizationGraph.scss';

export interface PipelineVisualizationGraphProps {
  graph: PipelineVisualizationTaskItem[][];
  namespace: string;
}

export const PipelineVisualizationGraph: React.FC<PipelineVisualizationGraphProps> = ({
  graph,
  namespace,
}) => {
  return (
    <div className="odc-pipeline-vis-graph">
      <div className="odc-pipeline-vis-graph__stages">
        <div className="odc-pipeline-vis-graph__stage">
          <div className="odc-pipeline-vis-task is-input-node">
            <ChevronCircleRightIcon />
          </div>
        </div>
        {graph.map((stage) => {
          return (
            <div
              className={cx('odc-pipeline-vis-graph__stage', { 'is-parallel': stage.length > 1 })}
              key={stage.map((t) => t.taskRef.name).join(',')}
            >
              <ul className="odc-pipeline-vis-graph__stage-column">
                {stage.map((task) => {
                  return (
                    <PipelineVisualizationTask
                      key={task.taskRef.name}
                      task={task}
                      namespace={namespace}
                    />
                  );
                })}
              </ul>
            </div>
          );
        })}
      </div>
    </div>
  );
};
