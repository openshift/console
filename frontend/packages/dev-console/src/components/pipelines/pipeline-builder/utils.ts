import { history, resourcePathFromModel } from '@console/internal/components/utils';
import { apiVersionForModel, referenceForModel } from '@console/internal/module/k8s';
import { ClusterTaskModel, PipelineModel } from '../../../models';
import { Pipeline, PipelineResourceTask, PipelineTask } from '../../../utils/pipeline-augment';
import { TASK_ERROR_STRINGS, TaskErrorType } from './const';
import { PipelineBuilderFormikValues, PipelineBuilderFormValues, TaskErrorMap } from './types';

export const getErrorMessage = (errorTypes: TaskErrorType[], errorMap: TaskErrorMap) => (
  taskName: string,
): string => {
  if (!taskName) {
    return TASK_ERROR_STRINGS[TaskErrorType.MISSING_NAME];
  }

  const errorList: TaskErrorType[] | undefined = errorMap?.[taskName];
  if (!errorList) return null;

  const hasRequestedError = errorList.filter((error) => errorTypes.includes(error));
  return hasRequestedError.length > 0 ? TASK_ERROR_STRINGS[hasRequestedError[0]] : null;
};

export const convertResourceToTask = (
  resource: PipelineResourceTask,
  runAfter?: string[],
): PipelineTask => {
  return {
    name: resource.metadata.name,
    runAfter,
    taskRef: {
      kind: resource.kind === ClusterTaskModel.kind ? ClusterTaskModel.kind : undefined,
      name: resource.metadata.name,
    },
    params: resource.spec.inputs?.params?.map((param) => ({
      name: param.name,
      value: param.default,
    })),
  };
};

export const getPipelineURL = (namespace: string) => {
  return `/k8s/ns/${namespace}/${referenceForModel(PipelineModel)}`;
};

const removeListRunAfters = (task: PipelineTask, listIds: string[]): PipelineTask => {
  if (task?.runAfter && listIds.length > 0) {
    // Trim out any runAfters pointing at list nodes
    const runAfter = (task.runAfter || []).filter(
      (runAfterName) => !listIds.includes(runAfterName),
    );

    return {
      ...task,
      runAfter,
    };
  }

  return task;
};

const removeEmptyDefaultParams = (task: PipelineTask): PipelineTask => {
  if (task.params?.length > 0) {
    // Since we can submit, this param has a default; check for empty values and remove
    return {
      ...task,
      params: task.params.filter((param) => !!param.value),
    };
  }

  return task;
};

export const convertBuilderFormToPipeline = (
  formValues: PipelineBuilderFormikValues,
  namespace: string,
  existingPipeline: Pipeline = {},
): Pipeline => {
  const { name, resources, params, tasks, listTasks } = formValues;
  const listIds = listTasks.map((listTask) => listTask.name);

  return {
    ...existingPipeline,
    apiVersion: apiVersionForModel(PipelineModel),
    kind: PipelineModel.kind,
    metadata: {
      ...existingPipeline.metadata,
      name,
      namespace,
    },
    spec: {
      ...existingPipeline.spec,
      params,
      resources,
      tasks: tasks.map((task) => removeEmptyDefaultParams(removeListRunAfters(task, listIds))),
    },
  };
};

export const convertPipelineToBuilderForm = (pipeline: Pipeline): PipelineBuilderFormValues => {
  if (!pipeline) return null;

  const {
    metadata: { name },
    spec: { params = [], resources = [], tasks = [] },
  } = pipeline;

  return {
    name,
    params,
    resources,
    tasks,
    listTasks: [],
  };
};

export const goToYAML = (existingPipeline?: Pipeline, namespace?: string) => {
  history.push(
    existingPipeline
      ? `${resourcePathFromModel(
          PipelineModel,
          existingPipeline?.metadata?.name,
          existingPipeline?.metadata?.namespace,
        )}/yaml`
      : `${getPipelineURL(namespace)}/~new`,
  );
};
