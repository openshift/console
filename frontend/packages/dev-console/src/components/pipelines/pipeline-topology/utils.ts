import * as dagre from 'dagre';
import * as _ from 'lodash';
import { Pipeline, PipelineRun } from '../../../utils/pipeline-augment';
import { getPipelineTasks, PipelineVisualizationTaskItem } from '../../../utils/pipeline-utils';
import {
  NODE_HEIGHT,
  NodeType,
  NODE_WIDTH,
  AddNodeDirection,
  PipelineLayout,
  DAGRE_BUILDER_PROPS,
  DAGRE_VIEWER_PROPS,
} from './const';
import {
  PipelineEdgeModel,
  NodeCreator,
  NodeCreatorSetup,
  SpacerNodeModelData,
  TaskListNodeModelData,
  TaskNodeModelData,
  PipelineMixedNodeModel,
  PipelineTaskNodeModel,
  BuilderNodeModelData,
  PipelineRunAfterNodeModelData,
} from './types';

const createGenericNode: NodeCreatorSetup = (type, width?) => (name, data) => ({
  id: name,
  data,
  height: NODE_HEIGHT,
  width: width != null ? width : NODE_WIDTH,
  type,
});

// Node variations
export const createTaskNode: NodeCreator<TaskNodeModelData> = createGenericNode(NodeType.TASK_NODE);
export const createSpacerNode: NodeCreator<SpacerNodeModelData> = createGenericNode(
  NodeType.SPACER_NODE,
  0,
);
export const createTaskListNode: NodeCreator<TaskListNodeModelData> = createGenericNode(
  NodeType.TASK_LIST_NODE,
);
export const createBuilderNode: NodeCreator<BuilderNodeModelData> = createGenericNode(
  NodeType.BUILDER_NODE,
);

export const getNodeCreator = (type: NodeType): NodeCreator<PipelineRunAfterNodeModelData> => {
  switch (type) {
    case NodeType.TASK_LIST_NODE:
      return createTaskListNode;
    case NodeType.BUILDER_NODE:
      return createBuilderNode;
    case NodeType.SPACER_NODE:
      return createSpacerNode;
    case NodeType.TASK_NODE:
    default:
      return createTaskNode;
  }
};

export const handleParallelToParallelNodes = (
  nodes: PipelineMixedNodeModel[],
): PipelineMixedNodeModel[] => {
  type ParallelNodeReference = {
    node: PipelineTaskNodeModel;
    runAfter: string[];
    atIndex: number;
  };
  type ParallelNodeMap = {
    [id: string]: ParallelNodeReference[];
  };

  // Collect only multiple run-afters
  const multipleRunBeforeMap: ParallelNodeMap = nodes.reduce((acc, node, idx) => {
    const {
      data: {
        task: { runAfter },
      },
    } = node;
    if (runAfter && runAfter.length > 1) {
      const id: string = [...runAfter]
        .sort((a, b) => a.localeCompare(b))
        .reduce((str, ref) => `${str}|${ref}`);

      if (!Array.isArray(acc[id])) {
        acc[id] = [];
      }
      acc[id].push({
        node,
        runAfter,
        atIndex: idx,
      });
    }
    return acc;
  }, {} as ParallelNodeMap);

  // Trim out single occurrences
  const multiParallelToParallelList: ParallelNodeReference[][] = Object.values(
    multipleRunBeforeMap,
  ).filter((data: ParallelNodeReference[]) => data.length > 1);

  if (multiParallelToParallelList.length === 0) {
    // No parallel to parallel
    return nodes;
  }

  // Insert a spacer node between the multiple nodes on the sides of a parallel-to-parallel
  const newNodes: PipelineMixedNodeModel[] = [];
  const newRunAfterNodeMap: { [nodeId: string]: string[] } = {};
  multiParallelToParallelList.forEach((p2p: ParallelNodeReference[]) => {
    // All nodes in each array share their runAfters
    const { runAfter } = p2p[0];

    const names: string[] = p2p.map((p2pData) => p2pData.node.id);
    const parallelSpacerName = `parallel-${names.join('-')}`;

    names.forEach((p2pNodeId) => {
      if (!Array.isArray(newRunAfterNodeMap[p2pNodeId])) {
        newRunAfterNodeMap[p2pNodeId] = [];
      }
      newRunAfterNodeMap[p2pNodeId].push(parallelSpacerName);
    });

    newNodes.push(
      createSpacerNode(parallelSpacerName, {
        task: {
          name: parallelSpacerName,
          runAfter,
        },
      }),
    );
  });

  // Update all impacted nodes to point at the spacer node as the spacer points at their original runAfters
  nodes.forEach((node) => {
    const newRunAfters: string[] | undefined = newRunAfterNodeMap[node.id];
    if (newRunAfters && newRunAfters.length > 0) {
      const {
        data: { task },
        type,
      } = node;

      const createNode: NodeCreator<PipelineRunAfterNodeModelData> = getNodeCreator(type);

      // Recreate the node with the new runAfter pointing to the spacer node
      newNodes.push(
        createNode(node.id, {
          ...node.data,
          task: {
            ...task,
            runAfter: newRunAfters,
          },
        }),
      );
    } else {
      // Unaffected node, just carry it over
      newNodes.push(node);
    }
  });

  return newNodes;
};

export const tasksToNodes = (
  taskList: PipelineVisualizationTaskItem[],
  pipeline?: Pipeline,
  pipelineRun?: PipelineRun,
): PipelineMixedNodeModel[] => {
  const nodeList: PipelineTaskNodeModel[] = taskList.map((task) =>
    createTaskNode(task.name, {
      task,
      pipeline,
      pipelineRun,
    }),
  );

  return handleParallelToParallelNodes(nodeList);
};

export const tasksToBuilderNodes = (
  taskList: PipelineVisualizationTaskItem[],
  onAddNode: (task: PipelineVisualizationTaskItem, direction: AddNodeDirection) => void,
  onNodeSelection: (task: PipelineVisualizationTaskItem) => void,
  getError: (taskName: string) => string,
  selectedIds: string[],
): PipelineMixedNodeModel[] => {
  return taskList.map((task) => {
    return createBuilderNode(task.name, {
      error: getError(task.name),
      task,
      selected: selectedIds.includes(task.name),
      onNodeSelection: () => {
        onNodeSelection(task);
      },
      onAddNode: (direction: AddNodeDirection) => {
        onAddNode(task, direction);
      },
    });
  });
};

export const getEdgesFromNodes = (nodes: PipelineMixedNodeModel[]): PipelineEdgeModel[] =>
  _.flatten(
    nodes.map((node) => {
      const {
        data: {
          task: { name: target, runAfter = [] },
        },
      } = node;

      if (runAfter.length === 0) return null;

      return runAfter.map((source) => ({
        id: `${source}~to~${target}`,
        type: 'edge',
        source,
        target,
      }));
    }),
  ).filter((edgeList) => !!edgeList);

export const getTopologyNodesEdges = (
  pipeline: Pipeline,
  pipelineRun?: PipelineRun,
): { nodes: PipelineMixedNodeModel[]; edges: PipelineEdgeModel[] } => {
  const taskList: PipelineVisualizationTaskItem[] = _.flatten(
    getPipelineTasks(pipeline, pipelineRun),
  );
  const nodes: PipelineMixedNodeModel[] = tasksToNodes(taskList, pipeline, pipelineRun);
  const edges: PipelineEdgeModel[] = getEdgesFromNodes(nodes);

  return { nodes, edges };
};

export const getLayoutData = (layout: PipelineLayout): dagre.GraphLabel => {
  switch (layout) {
    case PipelineLayout.DAGRE_BUILDER:
      return DAGRE_BUILDER_PROPS;
    case PipelineLayout.DAGRE_VIEWER:
      return DAGRE_VIEWER_PROPS;
    default:
      return null;
  }
};
