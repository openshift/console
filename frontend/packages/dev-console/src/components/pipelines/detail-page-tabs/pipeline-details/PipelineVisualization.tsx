import * as React from 'react';
import { Alert } from '@patternfly/react-core';
import { Pipeline, PipelineRun } from '../../../../utils/pipeline-augment';
import PipelineTopologyGraph from '../../pipeline-topology/PipelineTopologyGraph';
import { getTopologyNodesEdges } from '../../pipeline-topology/utils';
import { PipelineLayout } from '../../pipeline-topology/const';

import './PipelineVisualization.scss';

interface PipelineTopologyVisualizationProps {
  pipeline: Pipeline;
  pipelineRun?: PipelineRun;
}

const PipelineVisualization: React.FC<PipelineTopologyVisualizationProps> = ({
  pipeline,
  pipelineRun,
}) => {
  const { nodes, edges } = getTopologyNodesEdges(pipeline, pipelineRun);

  if (nodes.length === 0 && edges.length === 0) {
    // Nothing to render
    // TODO: Confirm wording with UX; ODC-1860
    return <Alert variant="info" isInline title="This Pipeline has no tasks to visualize." />;
  }

  return (
    <div className="odc-pipeline-visualization">
      <PipelineTopologyGraph
        id={pipelineRun?.metadata?.name || pipeline.metadata.name}
        nodes={nodes}
        edges={edges}
        layout={PipelineLayout.DAGRE_VIEWER}
      />
    </div>
  );
};

export default PipelineVisualization;
