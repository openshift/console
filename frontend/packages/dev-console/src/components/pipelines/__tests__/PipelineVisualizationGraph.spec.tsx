import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import * as Renderer from 'react-test-renderer';
import { PipelineVisualizationGraph } from '../PipelineVisualizationGraph';
import { PipelineVisualizationProps } from '../PipelineVisualization';
import { mockPipelineGraph } from './pipeline-visualization-test-data';
import { getPipelineTasks } from '../../../utils/pipeline-utils';
import { mockPipeline } from './pipeline-mock';
import { mockPipelineRun } from './pipelinerun-mock';

jest.mock('react-dom', () => ({
  findDOMNode: () => ({}),
  createPortal: (node) => node,
}));

jest.mock('../PipelineVisualizationTask');
describe('PipelineVisualizationGraph', () => {
  const props = {
    namespace: 'test',
    graph: mockPipelineGraph,
  };
  let wrapper: ShallowWrapper<PipelineVisualizationProps>;
  beforeEach(() => {
    wrapper = shallow(<PipelineVisualizationGraph {...props} />);
  });

  it('renders a Pipeline visualization graph', () => {
    expect(wrapper.exists()).toBeTruthy();
  });

  it('should contain right number of stages', () => {
    const noOfStages = wrapper.find('.odc-pipeline-vis-graph__stage-column').length;

    expect(noOfStages).toEqual(mockPipelineGraph.length);
  });

  it('should match the previous pipeline snapshot', () => {
    const tree = Renderer.create(<PipelineVisualizationGraph {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the previous pipelineRun snapshot', () => {
    const graph = getPipelineTasks(mockPipeline, mockPipelineRun);
    const tree = Renderer.create(
      <PipelineVisualizationGraph namespace={props.namespace} graph={graph} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
