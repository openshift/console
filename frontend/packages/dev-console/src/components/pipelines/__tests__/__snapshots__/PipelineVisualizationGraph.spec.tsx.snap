// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelineVisualizationGraph should match the previous pipeline snapshot 1`] = `
<div
  className="odc-pipeline-vis-graph"
>
  <div
    className="odc-pipeline-vis-graph__stages"
  >
    <div
      className="odc-pipeline-vis-graph__stage"
    >
      <div
        className="odc-pipeline-vis-task is-input-node"
      >
        <svg
          aria-hidden={true}
          aria-labelledby={null}
          fill="currentColor"
          height="1em"
          role="img"
          style={
            Object {
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm113.9 231L234.4 103.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L285.1 256 183.5 357.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L369.9 273c9.4-9.4 9.4-24.6 0-34z"
            transform=""
          />
        </svg>
      </div>
    </div>
    <div
      className="odc-pipeline-vis-graph__stage"
    >
      <ul
        className="odc-pipeline-vis-graph__stage-column"
      >
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            start-app
          </span>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline-vis-graph__stage is-parallel"
    >
      <ul
        className="odc-pipeline-vis-graph__stage-column"
      >
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            test-app-1
          </span>
        </li>
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            test-app-2
          </span>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline-vis-graph__stage is-parallel"
    >
      <ul
        className="odc-pipeline-vis-graph__stage-column"
      >
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            build-image-1
          </span>
        </li>
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            build-image-2
          </span>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline-vis-graph__stage"
    >
      <ul
        className="odc-pipeline-vis-graph__stage-column"
      >
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            deploy
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PipelineVisualizationGraph should match the previous pipelineRun snapshot 1`] = `
<div
  className="odc-pipeline-vis-graph"
>
  <div
    className="odc-pipeline-vis-graph__stages"
  >
    <div
      className="odc-pipeline-vis-graph__stage"
    >
      <div
        className="odc-pipeline-vis-task is-input-node"
      >
        <svg
          aria-hidden={true}
          aria-labelledby={null}
          fill="currentColor"
          height="1em"
          role="img"
          style={
            Object {
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm113.9 231L234.4 103.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L285.1 256 183.5 357.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L369.9 273c9.4-9.4 9.4-24.6 0-34z"
            transform=""
          />
        </svg>
      </div>
    </div>
    <div
      className="odc-pipeline-vis-graph__stage"
    >
      <ul
        className="odc-pipeline-vis-graph__stage-column"
      >
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            build-skaffold-web
          </span>
          <span
            className="is-done"
          >
            Succeeded
          </span>
          <span>
            32s
          </span>
        </li>
      </ul>
    </div>
    <div
      className="odc-pipeline-vis-graph__stage"
    >
      <ul
        className="odc-pipeline-vis-graph__stage-column"
      >
        <li
          className="odc-pipeline-vis-task"
        >
          <span>
            deploy-web
          </span>
          <span
            className="is-done"
          >
            Succeeded
          </span>
          <span>
            32s
          </span>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
